src/check/constrain/generate/call.rs

```diff
@@ -206,88 +206,72 @@ fn call_parameters(
-                    node: Node::PropertyCall {
-                        instance: Box::from(instance.clone()),
-                        property: Box::from(property.clone()),
-                    },
-                },
-                &env.var_mappings,
-            ))?;
-            constr.add("call property", &instance, &access);
-            Ok((constr.clone(), env.clone()))
+            (constr.clone(), Expected::new(property.pos, &Field { name: lit.clone() }))
         }
         Node::FunctionCall { name, args } => {
-            let (mut constr, env) = gen_vec(args, env, ctx, constr)?;
-            let instance_exp = Expected::try_from((instance, &env.var_mappings))?;
-            let mut args_with_self: Vec<Expected> = vec![instance_exp];
-            args_with_self.append(
-                &mut args
-                    .iter()
-                    .map(|a| Expected::try_from((a, &env.var_mappings)))
-                    .collect::<Result<_, _>>()?,
-            );
-
-            let node = Node::PropertyCall {
-                instance: Box::from(instance.clone()),
-                property: Box::from(property.clone()),
-            };
-            let instance_exp = Expected::try_from((
-                &AST { pos: instance.pos.union(property.pos), node },
-                &env.var_mappings,
-            ))?;
-            let access = Expected::new(
-                property.pos,
-                &Access {
-                    entity: Box::new(Expected::try_from((instance, &env.var_mappings))?),
-                    name: Box::new(Expected::new(
-                        property.pos,
-                        &Function {
-                            name: StringName::try_from(name.deref())?,
-                            args: args_with_self,
-                        },
-                    )),
-                },
-            );
+            let (constr, _) = gen_vec(args, env, ctx, constr)?;
+            let args = vec![last_inst.clone()]
+                .iter()
+                .chain(args)
+                .map(|ast| Expected::try_from((ast, &env.var_mappings)))
+                .collect::<Result<_, _>>()?;
 
-            constr.add("call class function", &instance_exp, &access);
-            Ok((constr, env))
+            let function = Function { name: StringName::try_from(name)?, args };
+            (constr, Expected::new(property.pos, &function))
         }
 
-        _ => Err(vec![TypeErr::new(property.pos, "Expected property call")]),
-    }
+        _ => return Err(vec![TypeErr::new(property.pos, "Expected property call")]),
+    };
+
+    let entire_call_as_ast: AST = instance.iter().rfold(property.clone(), |acc, ast| {
+        let (instance, property) = (Box::from(ast.clone()), Box::from(acc));
+        AST::new(ast.pos, Node::PropertyCall { instance, property })
+    });
+    let entire_call_as_ast = Expected::try_from((&entire_call_as_ast, &env.var_mappings))?;
+
+    let ast_without_access = if instance.len() == 1 {
```

Comment:
Bit sad that we don't use a more elegant float but instead an ugly if statement.
But not sure there is another way to do this.